packages/sdk/src/messageBackend/goldsky.ts

```diff
@@ -0,0 +1,374 @@
+     * Prepares a URI that is used for fetching messages
+     *
+     * @returns uri
+     */
+    get uri(): string {
+        // return `https://${this.env}.goldsky.io/c/nomad/gql/v1/graphql`
+        return `https://api.goldsky.io/c/nomad/gql/v1/graphql`
+    }
+
+    /**
+     * Prepares a Dispatch event from backend's internal message representation
+     *
+     * @returns A Dispatch event assiciated with transaction (if any)
+     */
+    async getDispatch(tx: string): Promise<Dispatch | undefined> {
+        const m = await this.getMessageByTx(tx);
+
+        if (!m) return undefined;
+
+        return {
+            args: {
+                messageHash: m.message_hash,
+                leafIndex: BigNumber.from(m.leaf_index),
+                destinationAndNonce: BigNumber.from(m.destination_and_nonce),
+                committedRoot: m.committed_root,
+                message: m.message,
+            },
+            transactionHash: m.dispatch_tx
+        }
+    }
+
+    /**
+     * Stores message into internal cache
+     */
+    storeMessage(m: GoldSkyMessage) {
+        this.messageCache.set(m.message_hash, m);
+        this.dispatchTxToMessageHash.set(m.dispatch_tx, m.message_hash);
+    }
+
+    /**
+     * Get the message representation associated with this message (if any)
+     * by message hash
+     *
+     * @returns A message representation (if any)
+     */
+    async getMessage(messageHash: string): Promise<GoldSkyMessage | undefined> {
+        let m = this.messageCache.get(messageHash);
+        if (!m) {
+            m = await this.fetchMessage({
+                messageHash
+            });
+            if (m) {
+                this.storeMessage(m);
+            }
+        }
+
+        return m;
+    }
+
+    /**
+     * Get the message representation associated with this message (if any)
+     * by dispatch transaction
+     *
+     * @returns A message representation (if any)
+     */
+    async getMessageByTx(tx: string): Promise<GoldSkyMessage | undefined> {
+        let m: GoldSkyMessage | undefined;
+        let messageHash = this.dispatchTxToMessageHash.get(tx);
+        if (!messageHash) {
+            m = await this.fetchMessage({
+                transactionHash: tx
+            });
+            if (m) {
+                this.storeMessage(m);
+            }
+        }
+
+        return m;
+    }
+
+
+    /**
+     * Get the `Dispatch` transaction hash associated with this message (if any)
+     *
+     * @returns A dispatch tx (if any)
+     */
+    async dispatchTx(messageHash: string): Promise<string | undefined> {
+        let m = await this.getMessage(messageHash);
+        return m?.dispatch_tx;
+    }
+
+    /**
+     * Get the `Relay` transaction hash associated with this message (if any)
+     *
+     * @returns A relay tx (if any)
+     */
+    async relayTx(messageHash: string): Promise<string | undefined> {
+        let m = await this.getMessage(messageHash);
+        return m?.relay_tx;
```

Comment:
If the relay_tx exists, we should return it.  If not, we should fetch the message again in case there has been an update in state

Reply:
same with updateTx and processTx

Reply:
If we're concerned about over-usage, we can set a timestamp and only fetch after at least 1 minute or something

Reply:
Addressed in the new commit (without over-usage features in it, as a separate feature later).